superset/connectors/sqla/utils.py

```diff
@@ -119,3 +120,22 @@ def get_virtual_table_metadata(dataset: "SqlaTable") -> List[Dict[str, str]]:
     except Exception as ex:
         raise SupersetGenericDBErrorException(message=str(ex)) from ex
     return cols
+
+
+def allow_adhoc_subquery(raw_sql: str) -> bool:
+    # pylint: disable=import-outside-toplevel
+    from superset import is_feature_enabled
+
+    if is_feature_enabled("ALLOW_ADHOC_SUBQUERY"):
+        return True
+
+    statement = sqlparse.parse(raw_sql)[0]
+    if has_table_query(statement):
```

Comment:
I think it's safe to assume the ad-hoc expression will have a single statement, but we might want to err on the side of caution just in case and do:

```python
for statement in sqlparse.parse(raw_sql):
    if has_table_query(statement):
        raise ...
```